crates/pallet-domains/src/runtime_registry.rs

```diff
@@ -0,0 +1,267 @@
+//! Runtime registry for domains
+
+use crate::pallet::{NextRuntimeId, RuntimeRegistry};
+use crate::Config;
+use codec::{Decode, Encode};
+use frame_support::PalletError;
+use scale_info::TypeInfo;
+use sp_core::Hasher;
+use sp_domains::{RuntimeId, RuntimeType};
+use sp_std::vec::Vec;
+use sp_version::RuntimeVersion;
+
+/// Runtime specific errors
+#[derive(TypeInfo, Encode, Decode, PalletError, Debug, PartialEq)]
+pub enum Error {
+    FailedToExtractRuntimeVersion,
+    InvalidSpecName,
+    SpecVersionNeedsToIncrease,
+    MaxRuntimeId,
+    MissingRuntimeObject,
+    MaxRuntimeUpgrades,
+}
+
+#[derive(TypeInfo, Debug, Encode, Decode, Clone, PartialEq, Eq)]
+pub struct RuntimeObject<Number, Hash>
+where
+    Number: Encode + Decode + Clone,
+    Hash: Encode + Decode + Clone,
+{
+    pub runtime_name: Vec<u8>,
+    pub runtime_type: RuntimeType,
+    pub runtime_upgrades: u32,
+    pub hash: Hash,
+    pub code: Vec<u8>,
+    pub version: RuntimeVersion,
+    pub created_at: Number,
+    pub updated_at: Number,
+}
+
+/// Extracts the runtime version of the provided code.
+pub(crate) fn runtime_version(code: &[u8]) -> Result<RuntimeVersion, Error> {
+    sp_io::misc::runtime_version(code)
+        .and_then(|v| RuntimeVersion::decode(&mut &v[..]).ok())
+        .ok_or(Error::FailedToExtractRuntimeVersion)
+}
+
+/// Upgrades current runtime with new runtime.
+pub(crate) fn can_upgrade_code(
```

Comment:
If we submit a patch to the upstream `can_set_code()` in the system pallet, we may simply reuse `frame_system::Pallet::<T>::can_set_code(current_version, code)` here next time.

Reply:
Yes that make sense but for now I'll leave a TODO